### PR TITLE
Dummy robot demo

### PR DESCRIPTION
**Pull request by Karsten1987**
This is a PR for a robot demo which includes a map_server, joint_state_publisher, robot_state_publisher and a laser_scan_publisher. It basically resembles the rrbot. 

The following PRs are dependent on this:
- TF2: https://github.com/ros/geometry2/pull/203
- CI: https://github.com/ros2/ci/pull/34

The following new packages have to be compiled for this:
- robot state publisher: https://github.com/ros2/robot_state_publisher.git
- robot_model: https://github.com/ros2/robot_model.git
- orocos_kdl: https://github.com/ros2/orocos_kinematics_dynamics.git
- urdfdom: https://github.com/ros2/urdfdom.git
- urdfdom_headers: https://github.com/ros/urdfdom_headers.git
- tinyxml_vendor: (optional) https://github.com/ros2/tinyxml_vendor.git

Depending on the success for the chocolatey packages, we can also use the upstream version of `urdfdom*` and `tinyxml`  

Buldfarm Status:
- linux: [Build Status](http://ci.ros2.org/job/ci_linux/2317/)
- mac: [Build Status](http://ci.ros2.org/job/ci_osx/1773/) (unrelated test failure)
- windows: In progress, we need tinyxml and eigen3 as chocolate packages

### CODE REVIEW

**Comment by Karsten1987** [comment]
@ros2/team I leave this PR labeled as `in progress` due to open PRs, but ask for feedback already.

**Comment by mikaelarguedas** [comment]
same as https://github.com/ros2/ci/pull/34#issuecomment-285918447, should all the dependencies listed above be added to the new repos file and this demo target a different branch than master?

File: dummy_robot/dummy_robot_bringup/package.xml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  <name>dummy_robot_bringup</name>
+  <version>0.0.0</version>
+  <description>
+    dummy robot bringup
+  </description>
+  <maintainer email="karsten@osrfoundation.org">Karsten Knese</maintainer>
+  <license>Apache License 2.0</license>
+
+  <buildtool_depend>ament_cmake</buildtool_depend>
+
```
Comment (Karsten1987) [inline dummy_robot/dummy_robot_bringup/package.xml]: done

File: dummy_robot/dummy_sensors/src/dummy_laser.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+#include <chrono>
+#include <iostream>
+#include <memory>
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "sensor_msgs/msg/laser_scan.hpp"
+
+#define DEG2RAD M_PI / 180.0
```
Comment (Karsten1987) [inline dummy_robot/dummy_sensors/src/dummy_laser.cpp]: copied from sick_lms code.

File: dummy_robot/dummy_sensors/src/dummy_laser.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  auto counter = 0.0;
+  auto amplitude = 1;
+  auto distance = 0.0;
+  while (rclcpp::ok()) {
+    counter += 0.1;
+    distance = std::abs(amplitude * std::sin(counter));
+
+    for (size_t i = 0; i < msg->ranges.size(); ++i) {
+      msg->ranges[i] = distance;
```
Comment (Karsten1987) [inline dummy_robot/dummy_sensors/src/dummy_laser.cpp]: it publishes 0 right now and are not necessary for the rviz demo.

**Comment by Karsten1987** [comment]
I didn't use the new map_server as this code was written before williams PR was merged. I'll definitely plan on using it, have to create a dummy map. Do we have any tutorial like demo maps?

For the tutorial repo file, I am not sure. I don't know if we have a real policy on what goes in there and what not. I still think that the robot_state_publisher is a rather crucial ROS(1/2) package, which should be in master - IMHO.

**Comment by gerkey** [comment]
@Karsten1987 if you're looking for an example map, the old Willow map is often helpful:

https://github.com/turtlebot/turtlebot_apps/tree/indigo/turtlebot_navigation/maps

File: dummy_robot/dummy_sensors/src/dummy_laser.cpp (hunk 3 above)
Comment (mikaelarguedas) [inline dummy_robot/dummy_sensors/src/dummy_laser.cpp]: sounds good, in that case maybe we can get rid of / comment out L51

File: dummy_robot/dummy_sensors/src/dummy_laser.cpp (hunk 2 above)
Comment (mikaelarguedas) [inline dummy_robot/dummy_sensors/src/dummy_laser.cpp]: Ok, while not changing the macro structure, I still think we should import `math.h` in here to use `M_PI`

File: dummy_robot/dummy_map_server/src/dummy_map_server.cpp [hunk 4]
```diff
@@ -16,13 +16,11 @@
 #include <iostream>
 #include <memory>
 
-#include "rcl/rcl.h"
 #include "rclcpp/rclcpp.hpp"
+#include "rcl/rcl.h"
```
Comment (mikaelarguedas) [inline dummy_robot/dummy_map_server/src/dummy_map_server.cpp]: Nit: rcl before rclcpp was good. nav_msgs is the include I was referring to (that should be before the rcl* ones)

**Comment by Karsten1987** [comment]
Tinyxml and Eigen3 chocolate packages can be found here:
https://github.com/ros2/choco-packages/releases

**Comment by Karsten1987** [comment]
Winner Winner Chicken Dinner!

Linux: [Build Status](http://ci.ros2.org/job/ci_linux/2350/)
OSX: [Build Status](http://ci.ros2.org/job/ci_osx/1806/)
Windows: [Build Status](http://ci.ros2.org/job/ci_windows/2452/)

As mentioned in the previous comments, `TinyXML` and `Eigen3` can simply be installed from `brew` or `aptitude`, for Windows we provide the choco packages.
I ran the demo successfully on Windows and piped it to RViz via the bridge!

@ros2/team If somebody could review that please.

**Comment by mikaelarguedas** [comment]
:+1: :tada:

**Comment by Karsten1987** [comment]
The changes for the installation wikipages can be found here:

linux: https://github.com/ros2/ros2/wiki/Linux-Development-Setup/_compare/bd41aa5615a3f59010aaacacf94a24366c7966bf...8fdf92a642716874ec549efc2b71d39bb282d16b

osx: https://github.com/ros2/ros2/wiki/OSX-Development-Setup/_compare/ec78654805e790f7d5a65fdd11192006654f255f...61ea04cecead0538005264eb270b172ba08b6c4e

windows: https://github.com/ros2/ros2/wiki/Windows-Development-Setup/_compare/01f788d93d7c7cef5a1bf6196973e5a1a514d0cc...38159b4f08531297271440d6dd042f32bd575b38

**Comment by wjwwood** [comment]
For the Windows one, do you think it would be better to specify that you mean to download the two `.nuget` files from that link? Or you could just link directly to the download of both.

Otherwise lgtm, thanks!

**Comment by Karsten1987** [comment]
And finally the wiki page 

https://github.com/ros2/ros2/wiki/dummy-robot-demo

File: dummy_robot/dummy_robot_bringup/launch/dummy_robot_bringup.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+import os
+
+from launch import LaunchDescriptor
+from launch.launcher import DefaultLauncher
+
+file_path = os.path.dirname(os.path.realpath(__file__))
+
+
+def launch():
```
Comment (dirk-thomas) [inline dummy_robot/dummy_robot_bringup/launch/dummy_robot_bringup.py]: Wrong signature for a `launch` file (see #140).

**Pull request merged**